packages/headless/src/controllers/facet-manager/headless-facet-manager.ts

```diff
@@ -1,18 +1,74 @@
 import {SearchEngine} from '../../app/search-engine/search-engine';
+import {setDesiredCount} from '../../features/facets/automatic-facets/automatic-facets-actions';
+import {loadAutomaticFacetsActions} from '../../features/facets/automatic-facets/automatic-facets-actions-loader';
+import {FacetResponse} from '../../features/facets/facet-set/interfaces/response';
+import {loadReducerError} from '../../utils/errors';
 import {
   buildCoreFacetManager,
-  FacetManager,
-  FacetManagerState,
+  FacetManager as CoreFacetManager,
+  FacetManagerState as CoreFacetManagerState,
   FacetManagerPayload,
 } from '../core/facet-manager/headless-core-facet-manager';
 
-export type {FacetManagerState, FacetManagerPayload, FacetManager};
+export type {FacetManagerPayload, CoreFacetManager, CoreFacetManagerState};
+
+export interface FacetManagerState extends CoreFacetManagerState {
+  /**
+   * The list of automatic facet responses.
+   */
+  automaticFacets: FacetResponse[] | undefined;
+}
+
+export interface FacetManager extends CoreFacetManager {
+  /**
+   * The state of the FacetManager.
+   */
+  state: FacetManagerState;
+}
+
+export interface FacetManagerProps {
+  /**
+   * The desired count of automatic facets.
+   * The default value is 0.
+   */
+  desiredCount: number;
+}
 
 /**
  * Creates a `FacetManager` instance.
  *
  * @param engine - The headless engine.
+ * @param props - The facet manager props.
+ * @returns A `FacetManager` controller instance.
  */
-export function buildFacetManager(engine: SearchEngine): FacetManager {
-  return buildCoreFacetManager(engine);
+export function buildFacetManager(
+  engine: SearchEngine,
+  props?: FacetManagerProps
+): FacetManager {
+  if (props && !loadAutomaticFacetsActions(engine)) {
+    throw loadReducerError;
+  }
+  const {dispatch} = engine;
+  const core = buildCoreFacetManager(engine);
+  const getAutomaticFacets = () =>
+    engine.state.search.response.generateAutomaticFacets?.facets;
+
+  if (props) {
+    dispatch(setDesiredCount(props.desiredCount));
+  }
+
+  return {
+    ...core,
+    get state() {
+      const modifiedAutomaticFacets = getAutomaticFacets()?.map((facet) => ({
+        ...facet,
+        facetId: `generated_${facet.field}`,
+      }));
+
+      return {
+        ...core.state,
+        automaticFacets: modifiedAutomaticFacets,
+      };
+    },
```

Comment:
I would create an util `buildAutomaticFaceManager` or something that do that., and the in
`buildFacetManager` return a `{...core, ...automatic}`

Reply:
The code you selected is not entirely for the `automatic` feature. There is also the `...core.state` that needs to be returned in the state. I think extracting this in a function would be confusing and hard since the `get` part is necessary for constantly updating the state. I still improved the code a bit for better clarity.